alns/select/AlphaUCB.py

```diff
@@ -0,0 +1,137 @@
+
+    where :math:`T_a(t - 1)` is the number of times action :math:`a` has been
+    played, and :math:`\\bar r_a(t - 1)` is the average reward of action
+    :math:`a`, both in the first :math:`t - 1` iterations. See
+    :meth:`~alns.select.AlphaUCB.AlphaUCB.update` for details on how
+    :math:`\\bar r_a` is updated.
+
+    .. note::
+
+        The average reward :math:`\\bar r_a(0)` of each action :math:`a \\in A`
+        is initialised to 1. The ``scores`` list passed into the
+        :math:`\\alpha`-UCB scheme should be 'reasonable' with respect to this
+        default.
+
+    Parameters
+    ----------
+    scores
+        A list of four non-negative elements, representing the rewards when the
+        candidate solution results in a new global best (idx 0), is better than
+        the current solution (idx 1), the solution is accepted (idx 2), or
+        rejected (idx 3).
+    alpha
+        The :math:`\\alpha` parameter controls the width of the confidence
+        interval. Larger values force the algorithm to select inferior
+        operators more frequently, resulting in more exploration.
+        :math:`\\alpha` must be in [0, 1].
+    num_destroy
+        Number of destroy operators.
+    num_repair
+        Number of repair operators.
+    op_coupling
+        Optional matrix that indicates coupling between destroy and repair
+        operators. Entry (i, j) is 1 if destroy operator i can be used in
+        conjunction with repair operator j and 0 otherwise.
+
+    References
+    ----------
+    .. [1] Hendel, G. 2022. Adaptive large neighborhood search for mixed
+           integer programming. *Mathematical Programming Computation* 14:
+           185 – 221.
+    """
+
+    def __init__(
+        self,
+        scores: List[float],
+        alpha: float,
+        num_destroy: int,
+        num_repair: int,
+        op_coupling: Optional[np.ndarray] = None,
+    ):
+        super().__init__(num_destroy, num_repair, op_coupling)
+
+        if not (0 <= alpha <= 1):
+            raise ValueError(f"Alpha {alpha:} outside [0, 1] not understood.")
+
+        if any(score < 0 for score in scores):
+            raise ValueError("Negative scores are not understood.")
+
+        if len(scores) < 4:
+            # More than four is OK because we only use the first four.
+            raise ValueError(f"Expected four scores, found {len(scores)}")
+
+        self._scores = scores
+        self._alpha = alpha
+
+        self._avg_rewards = np.ones_like(self._op_coupling, dtype=float)
+        self._times = np.zeros_like(self._op_coupling, dtype=int)
+        self._iter = 0  # current iteration
+
+    @property
+    def scores(self) -> List[float]:
+        return self._scores
+
+    @property
+    def alpha(self) -> float:
+        return self._alpha
+
+    def __call__(self, rnd, best, curr):
+        """
+        Returns the (destroy, repair) operator pair that maximises the average
+        reward and exploration bonus.
```

Comment:
Suggestion: perhaps add a small note in the class docstring about the Q value being a sum of the avg reward and exploration bonus? The Q expression looked quite complex when I read it first, but knowing that its just avg reward + exploration bonus immediately makes sense.